src/library_pthread.js

```diff
@@ -0,0 +1,655 @@
+    } else {
+      // We are the main thread, so we have the pthread warmup pool in this thread and can fire off JS thread creation
+      // directly ourselves.
+      __spawn_thread(threadParams);
+    }
+
+    return 0;
+  },
+
+  // TODO HACK! Remove this function, it is a JS side copy of the function pthread_testcancel() in library_pthread.c.
+  // Just call pthread_testcancel() everywhere.
+  _pthread_testcancel_js: function() {
+    if (!ENVIRONMENT_IS_PTHREAD) return;
+    if (!threadBlock) return;
+    var cancelDisabled = Atomics.load(HEAPU32, (threadBlock + {{{ C_STRUCTS.pthread.canceldisable }}} ) >> 2);
+    if (cancelDisabled) return;
+    var canceled = Atomics.load(HEAPU32, (threadBlock + {{{ C_STRUCTS.pthread.threadStatus }}} ) >> 2);
+    if (canceled == 2) throw 'Canceled!';
+  },
+
+  pthread_join__deps: ['_cleanup_thread', '_pthread_testcancel_js'],
+  pthread_join: function(thread, status) {
+    if (!thread) {
+      Module['printErr']('pthread_join attempted on a null thread pointer!');
+      return ERRNO_CODES.ESRCH;
+    }
+    if (ENVIRONMENT_IS_PTHREAD && selfThreadId == thread) {
+      Module['printErr']('PThread ' + thread + ' is attempting to join to itself!');
+      return ERRNO_CODES.EDEADLK;
+    }
+    else if (!ENVIRONMENT_IS_PTHREAD && PThread.mainThreadBlock == thread) {
+      Module['printErr']('Main thread ' + thread + ' is attempting to join to itself!');
+      return ERRNO_CODES.EDEADLK;
+    }
+    var self = {{{ makeGetValue('thread', C_STRUCTS.pthread.self, 'i32') }}};
+    if (self != thread) {
+      Module['printErr']('pthread_join attempted on thread ' + thread + ', which does not point to a valid thread, or does not exist anymore!');
+      return ERRNO_CODES.ESRCH;
+    }
+
+    var detached = Atomics.load(HEAPU32, (thread + {{{ C_STRUCTS.pthread.detached }}} ) >> 2);
+    if (detached) {
+      Module['printErr']('Attempted to join thread ' + thread + ', which was already detached!');
+      return ERRNO_CODES.EINVAL; // The thread is already detached, can no longer join it!
+    }
+    for(;;) {
+      var threadStatus = Atomics.load(HEAPU32, (thread + {{{ C_STRUCTS.pthread.threadStatus }}} ) >> 2);
+      if (threadStatus == 1) { // Exited?
+        var threadExitCode = Atomics.load(HEAPU32, (thread + {{{ C_STRUCTS.pthread.threadExitCode }}} ) >> 2);
+        if (status) {{{ makeSetValue('status', 0, 'threadExitCode', 'i32') }}};
+        Atomics.store(HEAPU32, (thread + {{{ C_STRUCTS.pthread.detached }}} ) >> 2, 1); // Mark the thread as detached.
+
+        if (!ENVIRONMENT_IS_WORKER) __cleanup_thread(thread);
+        else postMessage({ cmd: 'cleanupThread', thread: thread});
+        return 0;
+      }
+      // TODO HACK! Replace the _js variant with just _pthread_testcancel:
+      //_pthread_testcancel();
+      __pthread_testcancel_js();
+      // In Main runtime thread, assist pthreads in performing operations that they need to access the Emscripten main runtime for.
```

Comment:
main runtime thread is the JS VM thread that hosts the Emscripten runtime (the thread that initialized the heap and the C runtime). So in default build mode, this is the main browser thread, and in -s PROXY_TO_WORKER=1, it is a web worker thread.